testsuite/tests/typing-jkind-bounds/row.ml

```diff
@@ -0,0 +1,278 @@
+Error: This value is "aliased" but expected to be "unique".
+|}]
+
+let don't_cross_locality (x : int t @ local) = use_global x
+[%%expect{|
+Line 1, characters 58-59:
+1 | let don't_cross_locality (x : int t @ local) = use_global x
+                                                              ^
+Error: This value escapes its region.
+  Hint: This argument cannot be local,
+  because it is an argument in a tail call.
+|}]
+
+
+let cross_contention (x : int w @ contended) = use_uncontended x
+[%%expect{|
+val cross_contention : int w @ contended -> unit = <fun>
+|}]
+
+let don't_cross_portability (x : int w @ nonportable) = use_portable x
+[%%expect{|
+Line 1, characters 69-70:
+1 | let don't_cross_portability (x : int w @ nonportable) = use_portable x
+                                                                         ^
+Error: This value is "nonportable" but expected to be "portable".
+|}]
+
+(* Quality *)
+
+type 'a record : immutable_data with [ `A of 'a ] = { inner : 'a }
+[%%expect{|
+type 'a record = { inner : 'a; }
+|}]
+
+module type S = sig
+  type 'a polyvar = [ `A of 'a ]
+  type 'a record : immutable_data with 'a polyvar = { inner : 'a }
+end
+[%%expect{|
+module type S =
+  sig type 'a polyvar = [ `A of 'a ] type 'a record = { inner : 'a; } end
+|}]
+
+(* This is why we're able to give closed polymorphic variants best kinds *)
+module type S2 = S with type 'a polyvar = [ `B of int ref ]
+[%%expect{|
+Line 1, characters 17-59:
+1 | module type S2 = S with type 'a polyvar = [ `B of int ref ]
+                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: In this "with" constraint, the new definition of "polyvar"
+       does not match its original definition in the constrained signature:
+       Type declarations do not match:
+         type 'a polyvar = [ `B of int ref ]
+       is not included in
+         type 'a polyvar = [ `A of 'a ]
+       The type "[ `B of int ref ]" is not equal to the type "[ `A of 'a ]"
+       The second variant type does not allow tag(s) "`B"
+|}]
+
+(* When we give open polymorphic variants more precise kinds, we should make sure to give them not-best quality *)
+module type S = sig
+  type 'a polyvar = private [< `A of 'a | `B of int ref ]
+  type 'a record : immutable_data with 'a polyvar = { inner : 'a }
+end
+module type S2 = S with type 'a polyvar = [ `A of 'a ]
+module F(M : S2) = struct
+  let cross (x : int M.record @ contended) = use_uncontended x
+end
+(* CR layouts v2.8: This should be accepted *)
+[%%expect{|
+module type S =
+  sig
+    type 'a polyvar = private [< `A of 'a | `B of int ref ]
+    type 'a record = { inner : 'a; }
+  end
+module type S2 =
+  sig type 'a polyvar = [ `A of 'a ] type 'a record = { inner : 'a; } end
+module F :
+  functor (M : S2) -> sig val cross : int M.record @ contended -> unit end
+|}]
+
+
+(* Harder cases: row variables *)
+
+(* CR layouts v2.8: These are both correct, but we could probably infer a more precise kind for both. *)
```

Comment:
Not clear whether we should even try to do better here. I mean, we could infer better kinds for the parameters, but I think that would probably make the error message worse.

Instead, I would love to see tests that could plausibly be accepted -- basically just replace the `immediate` here with `immutable_data` or even `immutable_data with 'a`.

Reply:
Done.